src/libraries/Native/Unix/System.Security.Cryptography.Native.Android/pal_x509.c

```diff
@@ -0,0 +1,458 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+#include "pal_x509.h"
+
+#include <assert.h>
+#include <stdbool.h>
+#include <string.h>
+
+#define INIT_LOCALS(name, ...) \
+    enum { __VA_ARGS__, count_##name }; \
+    jobject name[count_##name] = { 0 }; \
+
+#define RELEASE_LOCALS(name, env) \
+{ \
+    for (int i = 0; i < count_##name; ++i) \
+    { \
+        jobject local = name[i]; \
+        if (local != NULL) \
+            (*env)->DeleteLocalRef(env, local); \
+    } \
+} \
+
+#define VALUE_FAIL -1
+#define BUFFER_FAIL FAIL
+
+static int32_t PopulateByteArray(JNIEnv *env, jbyteArray source, uint8_t *dest, int32_t len);
+static int32_t PopulateString(JNIEnv *env, jstring source, char *dest, int32_t len);
+
+// Handles both DER and PEM formats
+jobject /*X509Certificate*/ AndroidCryptoNative_DecodeX509(const uint8_t *buf, int32_t len)
+{
+    assert(buf != NULL && len > 0);
+    JNIEnv* env = GetJNIEnv();
+
+    jobject ret = NULL;
+    INIT_LOCALS(loc, bytes, stream, certType, certFactory)
+
+    // byte[] bytes = new byte[] { ... }
+    // InputStream stream = new ByteArrayInputStream(bytes);
+    loc[bytes] = (*env)->NewByteArray(env, len);
+    (*env)->SetByteArrayRegion(env, loc[bytes], 0, len, (const jbyte*)buf);
```

Comment:
I'll admit I've mostly been skimming the JNI code, "reading" the Java through it, but it occurs to me to ask:

How is OOM handled?  e.g. what if the call to NewByteArray failed because it couldn't get a big enough array?  Will that just return null with the "pending exception" bit set (so it should be followed with `ON_EXCEPTION_PRINT_AND_GOTO(cleanup)`), or something else?  And then what would happen on this line?

Reply:
Ah, yes - all the `New*` can return null with a pending OutOfMemory exception and need the exception check.